### PR TITLE
fix(backends/lean): add a significant amount of keywords

### LINKED ISSUE #139: Usage of `as` in Lean can cause syntactical errors
Consider:

```lean
/- [avl_verification::{avl_verification::AVLTreeSet<T>#2}::insert_phase1]:
   Source: 'src/main.rs', lines 199:4-199:49 -/
def AVLTreeSet.insert_phase1
  (T : Type) (OrdInst : Ord T) (self : AVLTreeSet T) (value : T) :
  Result (Bool × (AVLTreeSet T))
  :=
  do
  let (b, as) ← AVLTreeSet.insert_phase1_loop T OrdInst value self.root
  Result.ok (b, { root := as })
```

`as` seems to be a keyword or something causing:
```
▶ 280:10-280:13: error:
unexpected token 'as'; expected term
```

Renaming `as` to `u` fixes the issue.

### PR DESCRIPTION
**Pull request by RaitoBezarius**
This started off from the fact that `as` is a reserved keyword and cannot be used as a variable name.
And this ended up adding many keywords.

Fixes #139.

cc @sonmarcho

### CODE REVIEW

**Comment by sonmarcho** [comment]
Can you also add "from" as a keyword?

**Comment by sonmarcho** [comment]
> If we're in the process of fixing Lean keyword issues, should we complete the list using, e.g., https://github.com/leanprover/lean4/blob/master/doc/latex/lstlean.tex#L12 ?

Yes, that's a good idea. I was not aware of this list.

**Comment by RaitoBezarius** [comment]
I'm doing it.

**Comment by RaitoBezarius** [comment]
Done @sonmarcho @R1kM PTAL.

File: compiler/ExtractBase.ml [hunk 1]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "attribute";
+          "axiom";
+          "axioms";
+          "begin";
+          "break";
           "by";
+          "calc";
+          "cases";
+          "catch";
+          "check";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:940]: I think we can remove this one

File: compiler/ExtractBase.ml [hunk 2]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
           "class";
+          "clear";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:942]: This is a tactic? I think we can remove it

File: compiler/ExtractBase.ml [hunk 3]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
         [
+          "Pi";
+          "Prop";
+          "Sort";
+          "Type";
+          "abbrev";
+          "add_key_equivalence";
+          "alias";
+          "allGoals";
+          "apply";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:926]: remove allGoals and apply: those are tactics

File: compiler/ExtractBase.ml [hunk 4]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "as";
+          "assume";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:928]: remove

File: compiler/ExtractBase.ml [hunk 5]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "catch";
+          "check";
           "class";
+          "clear";
+          "const";
+          "constant";
+          "constants";
+          "constructor";
+          "continue";
+          "contradiction";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:948]: remove

File: compiler/ExtractBase.ml [hunk 6]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "declare_syntax_cat";
+          "declare_trace";
           "decreasing_by";
           "def";
+          "definition";
           "deriving";
+          "dif";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:955]: remove

File: compiler/ExtractBase.ml [hunk 7]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
           "do";
+          "done";
           "else";
           "end";
+          "eval";
+          "exact";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:961]: remove those two

File: compiler/ExtractBase.ml [hunk 8]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "example";
+          "exists";
+          "export";
+          "extends";
           "for";
+          "forall";
+          "from";
+          "fun";
+          "generalize";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:970]: remove

File: compiler/ExtractBase.ml [hunk 9]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
           "have";
+          "help";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:972]: remove

File: compiler/ExtractBase.ml [hunk 10]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "precedence";
+          "prefix";
+          "prelude";
+          "print";
+          "private";
+          "protected";
+          "raw";
+          "record";
+          "reduce";
+          "refine";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:1017]: remove

File: compiler/ExtractBase.ml [hunk 11]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "simp_all";
           "structure";
+          "suffices";
           "syntax";
           "termination_by";
           "then";
-          "Type";
+          "theorem";
+          "theory";
+          "this";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:1040]: remove

**Review by sonmarcho — CHANGES_REQUESTED** [review]
There are several keywords which are tactics, or which require to be prefixed with `#` (I think), so it should be ok to remove them.

File: compiler/ExtractBase.ml [hunk 12]
```diff
@@ -915,38 +915,140 @@ let keywords () =
...
+          "raw";
+          "record";
+          "reduce";
+          "refine";
+          "register_simp_ext";
+          "renaming";
+          "repeat";
+          "replacing";
+          "reserve";
+          "return";
```
Comment (sonmarcho) [inline compiler/ExtractBase.ml:1023]: Remove this one: I'm actually using it

**Comment by RaitoBezarius** [comment]
> There are several keywords which are tactics, or which require to be prefixed with `#` (I think), so it should be ok to remove them.

I removed all the tactics I could find, PTAL.

**Comment by sonmarcho** [comment]
I think we can also remove `check` and `help` because they are commands (which must be prefixed by `#`)?

**Comment by RaitoBezarius** [comment]
> I think we can also remove `check` and `help` because they are commands (which must be prefixed by `#`)?

Apologies, I must have missed it, this is done now.

**Review by sonmarcho — APPROVED** [review]

**Pull request merged**